filecoin/eventrecorder/eventrecorder.go

```diff
@@ -0,0 +1,137 @@
+package eventrecorder
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/application-research/autoretrieve/filecoin"
+	"github.com/application-research/filclient/rep"
+	"github.com/filecoin-project/go-fil-markets/retrievalmarket"
+	"github.com/google/uuid"
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+var _ filecoin.RetrievalEventListener = (*EventRecorder)(nil)
+
+var log = logging.Logger("eventrecorder")
+
+func NewEventRecorder(instanceId string, endpointURL string) *EventRecorder {
+	return &EventRecorder{instanceId, fmt.Sprintf("%s/retrieval-events", endpointURL)}
+}
+
+type EventRecorder struct {
+	instanceId  string
+	endpointURL string
+}
+
+type eventReport struct {
+	RetrievalId       uuid.UUID              `json:"retrievalId"`
+	InstanceId        string                 `json:"instanceId"`
+	Cid               string                 `json:"cid"`
+	StorageProviderId peer.ID                `json:"storageProviderId"`
+	Phase             rep.Phase              `json:"phase"`
+	PhaseStartTime    time.Time              `json:"phaseStartTime"`
+	Event             rep.RetrievalEventCode `json:"event"`
+	EventTime         time.Time              `json:"eventTime"`
+	EventDetails      interface{}            `json:"eventDetails"`
+}
+
+type eventDetailsSuccess struct {
+	ReceivedSize uint64 `json:"receivedSize,omitempty"`
+}
+
+type eventDetailsError struct {
+	Error string `json:"error,omitempty"`
+}
+
+func (er EventRecorder) newEventReport(
```

Comment:
is this function getting us much?

Reply:
Yeah, I'm simply using Go's strict function argument requirement to ensure that I have all the properties set on the struct I'm making, because I make it in a bunch of places I wanted to avoid the possibility of leaving one off or some other inconsistency. Do you have a suggestion of an alternative pattern to deal with this?

Reply:
I mean doesn't abbrieviated struct initialization notation itself -- as you're using in the function body here -- accomplish the same thing?